### PR TITLE
[FEATURE] Added default Rendertron support

### PR DESCRIPTION
**Pull request by lewisvoncken**
## Description

Many frontend frameworks rely on JavaScript to show content. This can mean Google might take some time to index your content or update the indexed content. One workaround option is to implement dynamic rendering. Dynamic rendering means switching between client-side rendered and pre-rendered content for specific user agents. This codelab will walk you through implementing dynamic rendering using Rendertron, which is an open source solution based on headless Chromium.

source: https://codelabs.developers.google.com/codelabs/dynamic-rendering#0

With 3 new environment variables you can easily setup Server Side Rendering with Rendertron.

- RENDERTRON_URL
*Specify the Base URL of the Rendertron proxy service.*
- RENDERTRON_USER_AGENTS
*Regular expression to match user agent to proxy. Defaults to a set of bots that do not perform well with pages that require JavaScript.*
- RENDERTRON_EXCLUDE_URL_PATTERN
*Regular expression used to exclude request URL paths. Defaults to a set of typical static asset file extensions.*

It is possible to self-host Rendertron or even refer to SeoSnap (https://seosnap.io/)

### Verification Steps

1. Add the following lines to your .env (**important:** in this example i use the appspot rendertron which shouldn't be used in production)
```
RENDERTRON_URL=https://localhost:3000/render
RENDERTRON_USER_AGENTS="PreCacher|SeoScan|SitemapCacheWarmer|Screaming|Mojolicious|googlebot|bingbot|yandex|Whatsapp|baiduspider|twitterbot|facebookexternalhit|rogerbot|linkedinbot|embedly|quora link preview|showyoubot|outbrain|pinterest|slackbot|vkShare|W3C_Validator|Lighthouse"
RENDERTRON_EXCLUDE_URL_PATTERN="graphql|\.(js|css|xml|less|png|jpg|jpeg|gif|pdf|doc|txt|ico|rss|zip|mp3|rar|exe|wmv|doc|avi|ppt|mpg|mpeg|tif|wav|mov|psd|ai|xls|mp4|m4a|swf|dat|dmg|iso|flv|m4v|torrent|ttf|woff|svg|eot)"
```
2. Install Rendertron locally and run it on port 3000 or use https://render-tron.appspot.com/
3. Switch the user-agent to for example googlebot and you should get a fully rendered page
4. It is possible that you have to set `closeBrowser` in the rendertron configuration to true

image

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.

### Resolved issues:
1. [x] resolves magento/pwa-studio#3166: [FEATURE] Added default Rendertron support

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -80,7 +80,8 @@
   },
   "optionalDependencies": {
     "bundlesize": "~0.17.1",
-    "sharp": "~0.23.3"
+    "sharp": "~0.23.3",
+    "rendertron-middleware": "^0.1.5"
```
Comment (jimbo) [inline package.json:84]:
```suggestion
    "rendertron-middleware": "~0.1.5",
    "sharp": "~0.23.3"
```

File: packages/upward-js/lib/createUpwardServer.js [hunk 2]
```diff
@@ -43,10 +43,25 @@ async function createUpwardServer({
     env = process.env,
     before = () => {}
 }) {
+    logUrl = true;
```
Comment (jimbo) [inline packages/upward-js/lib/createUpwardServer.js:46]: Left in by accident, maybe?

```suggestion
```

**Review by jimbo — COMMENTED** [review]
Very nice; thanks for your contribution, Lewis. Seems very straightforward, so I'd love to try this out.

File: packages/upward-js/lib/createUpwardServer.js (hunk 2 above)
Comment (lewisvoncken) [reply packages/upward-js/lib/createUpwardServer.js]: @jimbo I have updated the PR

File: package.json (hunk 1 above)
Comment (lewisvoncken) [reply package.json]: @jimbo I have updated the PR

**Comment by lewisvoncken** [comment]
@jimbo did you had time to check this PR?

**Comment by sirugh** [comment]
@magento create issue from PR

**Comment by sirugh** [comment]
@lewisvoncken Can you help me understand the dependency order of this PR + #3090 + https://github.com/GoogleChrome/rendertron/pull/654?

It's not clear to me which needs to be merged first or if this is all dependent on the rendertron PR. Thanks!

**Comment by tkacheva** [comment]
@lewisvoncken thank you for submitting a PR! It took us a while to review it internally as we recognize the value Rendition can add to the product, but at the same time we want to make sure the pre-rendering mechanism added to the product works across all deployment models (on-prem and cloud). We start investigating the options internally and welcome you to contribute. Let me know if you are interested

**Pull request closed without merging**